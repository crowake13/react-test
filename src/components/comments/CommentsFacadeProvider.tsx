import React from 'react';
import * as RENDER_LOG from '../../constants/render-log';
import { CommentsContext } from '../../stores/comments/comments.context';
import { facade, ICommentsFacade } from '../../stores/comments/comments.facade';
import withRenderLog from '../shared/withRenderLog';

interface CommentsFacadeProviderProps extends Partial<ICommentsFacade> {
  children: any;
}

const CommentsFacadeProvider = ({
  children,
  ...props
}: CommentsFacadeProviderProps) => {
  return (
    <CommentsContext.Provider
      value={{
        errors: props.errors ?? facade.errors,
        error$: props.error$ ?? facade.error$,
        active$: props.active$ ?? facade.active$,
        activate: props.activate ?? facade.activate,
        getAll: props.getAll ?? facade.getAll,
        getById: props.getById ?? facade.getById,
        hashMap: props.hashMap ?? facade.hashMap,
        entitie$: props.entitie$ ?? facade.entitie$,
        isFetchingBySlug: props.isFetchingBySlug ?? facade.isFetchingBySlug,
        isFetching: props.isFetching ?? facade.isFetching,
        isFetching$: props.isFetching$ ?? facade.isFetching$,
        fetch: props.fetch ?? facade.fetch,
        fetchAll: props.fetchAll ?? facade.fetchAll
      }}
    >
      {children}
    </CommentsContext.Provider>
  );
};

export default withRenderLog({ greeting: RENDER_LOG.GREETING })(
  CommentsFacadeProvider
);
