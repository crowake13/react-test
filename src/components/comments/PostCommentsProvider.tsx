import React, { useState } from 'react';
import * as RENDER_LOG from '../../constants/render-log';
import { useComments } from '../../hooks/comments/comments.hook';
import PostCommentsContext from '../../stores/comments/post-comments.context';
import { ID } from '../../stores/entities/entity.facade';
import withRenderLog from '../shared/withRenderLog';

interface PostCommentsProvider {
  postId: ID;
  areCommentsVisibleOnInit?: boolean;
  children: any;
}

const PostCommentsProvider = ({
  postId,
  areCommentsVisibleOnInit,
  children
}: PostCommentsProvider) => {
  const [areCommentsVisible, toggleCommentsVisibility] = useState(
    areCommentsVisibleOnInit ?? false
  );

  const [, getCommentsByPostId] = useComments();

  return (
    <PostCommentsContext.Provider
      value={{
        comments: getCommentsByPostId(postId),
        areCommentsVisible: areCommentsVisible,
        toggleCommentsVisibility: toggleCommentsVisibility
      }}
    >
      {children}
    </PostCommentsContext.Provider>
  );
};

export default withRenderLog({ greeting: RENDER_LOG.GREETING })(
  PostCommentsProvider
);
