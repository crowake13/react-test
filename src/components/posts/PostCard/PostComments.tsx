import React, { useContext } from 'react';
import * as RENDER_LOG from '../../../constants/render-log';
import PostCommentsContext from '../../../stores/comments/post-comments.context';
import { CommentList } from '../../comments/CommentsList';
import withRenderLog from '../../shared/withRenderLog';

interface PostCommentsProps {
  loadingCommentsLabel: string;
  noCommentsLabel: string;
}

const PostComments = ({
  loadingCommentsLabel,
  noCommentsLabel
}: PostCommentsProps) => {
  const { areCommentsVisible, comments, toggleCommentsVisibility } = useContext(
    PostCommentsContext
  );

  const headerTitle = !comments
    ? loadingCommentsLabel
    : !comments.length
    ? noCommentsLabel
    : `Comments (${comments.length})`;

  return (
    <div
      className={`post-comments-container bg-light border-top ${
        areCommentsVisible ? 'full-card-height' : ''
      }`}
    >
      <div className="card-header">
        <div title={headerTitle} className="card-header-title">
          {headerTitle}
        </div>
        {!comments || !comments.length ? null : (
          <button
            type="button"
            className="close"
            aria-label="Close"
            onClick={() => toggleCommentsVisibility(!areCommentsVisible)}
          >
            <span aria-hidden="true">
              {areCommentsVisible ? (
                <svg
                  className="bi bi-chevron-down"
                  width="1em"
                  height="1em"
                  viewBox="0 0 20 20"
                  fill="currentColor"
                  xmlns="http://www.w3.org/2000/svg"
                >
                  <path
                    fillRule="evenodd"
                    d="M3.646 6.646a.5.5 0 01.708 0L10 12.293l5.646-5.647a.5.5 0 01.708.708l-6 6a.5.5 0 01-.708 0l-6-6a.5.5 0 010-.708z"
                    clipRule="evenodd"
                  ></path>
                </svg>
              ) : (
                <svg
                  className="bi bi-chevron-up"
                  width="1em"
                  height="1em"
                  viewBox="0 0 20 20"
                  fill="currentColor"
                  xmlns="http://www.w3.org/2000/svg"
                >
                  <path
                    fillRule="evenodd"
                    d="M9.646 6.646a.5.5 0 01.708 0l6 6a.5.5 0 01-.708.708L10 7.707l-5.646 5.647a.5.5 0 01-.708-.708l6-6z"
                    clipRule="evenodd"
                  ></path>
                </svg>
              )}
            </span>
          </button>
        )}
      </div>
      {areCommentsVisible && comments && comments.length ? (
        <CommentList comments={comments} />
      ) : null}
    </div>
  );
};

export default withRenderLog({ greeting: RENDER_LOG.GREETING })(PostComments);
