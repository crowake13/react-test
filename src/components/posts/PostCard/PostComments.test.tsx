import React from 'react';
import { render, unmountComponentAtNode } from 'react-dom';
import { act } from 'react-dom/test-utils';
import { of } from 'rxjs';
import CommentsFacadeProvider from '../../comments/CommentsFacadeProvider';
import PostCommentsProvider from '../../comments/PostCommentsProvider';
import PostComments from './PostComments';

let container: HTMLElement | null = null;

beforeEach(() => {
  // setup a DOM element as a render target
  container = document.createElement('div');
  document.body.appendChild(container);
});

afterEach(() => {
  if (!container) {
    return;
  }

  // cleanup on exiting
  unmountComponentAtNode(container);
  container.remove();
  container = null;
});

const postId = 1;
const comments = [
  {
    postId,
    id: 1,
    name: 'title 1.1',
    email: 'email 1.1',
    body: 'body 1.1'
  },
  {
    postId,
    id: 2,
    name: 'title 1.2',
    email: 'email 1.2',
    body: 'body 1.2'
  }
];

it('renders comments on init', () => {
  act(() => {
    render(
      <CommentsFacadeProvider entitie$={of(comments)}>
        <PostCommentsProvider postId={postId} areCommentsVisibleOnInit={true}>
          <PostComments
            loadingCommentsLabel="Comments are loading..."
            noCommentsLabel="There are no comments"
          />
        </PostCommentsProvider>
      </CommentsFacadeProvider>,
      container
    );
  });
  expect(container?.innerHTML).toMatchSnapshot();
});

it('renders comments after clicking visibility toggle', () => {
  act(() => {
    render(
      <CommentsFacadeProvider entitie$={of(comments)}>
        <PostCommentsProvider postId={postId}>
          <PostComments
            loadingCommentsLabel="Comments are loading..."
            noCommentsLabel="There are no comments"
          />
        </PostCommentsProvider>
      </CommentsFacadeProvider>,
      container
    );
  });
  expect(container?.innerHTML).toMatchSnapshot();

  const button: HTMLButtonElement | null =
    container?.querySelector('button.close') ?? null;

  act(() => {
    button?.click();
  });
  expect(
    container
      ?.querySelector('.post-comments-container')
      ?.classList.contains('full-card-height')
  ).toBe(true);
  expect(container?.innerHTML).toMatchSnapshot();
});
