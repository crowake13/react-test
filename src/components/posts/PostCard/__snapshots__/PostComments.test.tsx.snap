// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`renders comments after clicking visibility toggle 1`] = `"<div class=\\"bg-light border-top \\"><div class=\\"card-header\\"><span>Comments (2)</span><button type=\\"button\\" class=\\"close float-right\\" aria-label=\\"Close\\"><span aria-hidden=\\"true\\"><svg class=\\"bi bi-chevron-up\\" width=\\"1em\\" height=\\"1em\\" viewBox=\\"0 0 20 20\\" fill=\\"currentColor\\" xmlns=\\"http://www.w3.org/2000/svg\\"><path fill-rule=\\"evenodd\\" d=\\"M9.646 6.646a.5.5 0 01.708 0l6 6a.5.5 0 01-.708.708L10 7.707l-5.646 5.647a.5.5 0 01-.708-.708l6-6z\\" clip-rule=\\"evenodd\\"></path></svg></span></button></div></div>"`;

exports[`renders comments after clicking visibility toggle 2`] = `"<div class=\\"bg-light border-top full-card-height\\"><div class=\\"card-header\\"><span>Comments (2)</span><button type=\\"button\\" class=\\"close float-right\\" aria-label=\\"Close\\"><span aria-hidden=\\"true\\"><svg class=\\"bi bi-chevron-down\\" width=\\"1em\\" height=\\"1em\\" viewBox=\\"0 0 20 20\\" fill=\\"currentColor\\" xmlns=\\"http://www.w3.org/2000/svg\\"><path fill-rule=\\"evenodd\\" d=\\"M3.646 6.646a.5.5 0 01.708 0L10 12.293l5.646-5.647a.5.5 0 01.708.708l-6 6a.5.5 0 01-.708 0l-6-6a.5.5 0 010-.708z\\" clip-rule=\\"evenodd\\"></path></svg></span></button></div><div class=\\"card-body m-0\\"><div class=\\"mb-2\\"><h6 class=\\"card-subtitle my-2 text-muted\\">email 1.1</h6><h5 class=\\"card-title\\">title 1.1</h5><p class=\\"card-text\\">body 1.1</p></div><div class=\\"border-top mb-2\\"><h6 class=\\"card-subtitle my-2 text-muted\\">email 1.2</h6><h5 class=\\"card-title\\">title 1.2</h5><p class=\\"card-text\\">body 1.2</p></div></div></div>"`;

exports[`renders comments on init 1`] = `"<div class=\\"bg-light border-top full-card-height\\"><div class=\\"card-header\\"><span>Comments (2)</span><button type=\\"button\\" class=\\"close float-right\\" aria-label=\\"Close\\"><span aria-hidden=\\"true\\"><svg class=\\"bi bi-chevron-down\\" width=\\"1em\\" height=\\"1em\\" viewBox=\\"0 0 20 20\\" fill=\\"currentColor\\" xmlns=\\"http://www.w3.org/2000/svg\\"><path fill-rule=\\"evenodd\\" d=\\"M3.646 6.646a.5.5 0 01.708 0L10 12.293l5.646-5.647a.5.5 0 01.708.708l-6 6a.5.5 0 01-.708 0l-6-6a.5.5 0 010-.708z\\" clip-rule=\\"evenodd\\"></path></svg></span></button></div><div class=\\"card-body m-0\\"><div class=\\"mb-2\\"><h6 class=\\"card-subtitle my-2 text-muted\\">email 1.1</h6><h5 class=\\"card-title\\">title 1.1</h5><p class=\\"card-text\\">body 1.1</p></div><div class=\\"border-top mb-2\\"><h6 class=\\"card-subtitle my-2 text-muted\\">email 1.2</h6><h5 class=\\"card-title\\">title 1.2</h5><p class=\\"card-text\\">body 1.2</p></div></div></div>"`;
